lib/land/trackers/api_tracker.rb

```diff
@@ -206,24 +199,70 @@ def user_agent
                      Land.config.blank_user_agent_string
 
         @user_agent = UserAgent[user_agent]
+        @user_agent.user_agent_type = UserAgentType['user']
 
         if Land.config.identify_crawlers && defined?(CrawlerDetect)
           crawler_detect = CrawlerDetect.new(user_agent)
-          user_agent_type = crawler_detect.is_crawler? ? 'crawl' : 'api'
-          @user_agent.update(user_agent_type: UserAgentType[user_agent_type])
+          @user_agent.user_agent_type = UserAgentType['crawl'] if crawler_detect.is_crawler?
         end
 
         browser = ::Browser.new(user_agent)
+        land_browser = Land::Browser[browser.name]
 
-        @user_agent.update(
-          browser: Browser[browser.name],
-          device: Device[browser.device.name],
-          platform: Platform[browser.platform.name],
-          browser_version: browser.version
-        )
+        @user_agent.browser = land_browser
+        @user_agent.device = Device[browser.device.name]
+        @user_agent.platform = Platform[browser.platform.name]
+        @user_agent.browser_version = browser.version
+        @user_agent.device_resolution = device_resolution
+        @user_agent.browser_color_preference = browser_color_preference
+
+        @user_agent.save! if @user_agent.changed?
+
+        @user_agent
+      rescue ActiveRecord::RecordNotUnique
+        retry
+      rescue ActiveRecord::RecordInvalid => e
+        retry if e.message == 'Validation failed: User agent has already been taken'
 
+        add_error_tag_to_land_span(e)
         @user_agent
       end
+
+      def device_resolution
+        device_width = params.dig('device_resolution', 'width')
+        device_height = params.dig('device_resolution', 'height')
+        device_orientation = params.dig('device_resolution', 'orientation')
+      
+        return unless device_width && device_height && device_orientation
+      
+        DeviceResolution.find_or_create_by(
+          device_resolution: "#{device_width}x#{device_height}",
+          width: device_width,
+          height: device_height,
+          orientation: device_orientation
+        )
+      end
+      
+      def browser_color_preference
+        dark_mode = params.dig('color_scheme_preference', 'is_dark_mode')
+        light_mode = params.dig('color_scheme_preference', 'is_light_mode')
+        no_preference = params.dig('color_scheme_preference', 'is_no_preference')
+      
+        return unless dark_mode && light_mode && no_preference
+      
+        BrowserColorPreference.find_or_create_by(
```

Comment:
Please pay attention to comments, make sure we are reading them thoroughly. There should only be one call that saves or updates here, the `UserAgent#save!` call. `find_or_create_by` does the creation, which is an additional database call. 

```suggestion
        BrowserColorPreference.find_or_initialize_by(
```